File: PracticaVariablesdeCondicion/EjemploCondition.py
import threading
import time

class EjemploCondition:
    def __init__(self):
        self.lock = threading.Lock()
        self.condicion = threading.Condition(self.lock)
        self.listo = False

    def esperar(self):
        with self.condicion:  # Adquirimos el lock y la condición
            while not self.listo:
                print(f"{threading.current_thread().name} esperando...")
                self.condicion.wait()  # Espera a que se cumpla la condición
            print(f"{threading.current_thread().name} ha continuado.")

    def senalar(self):
        with self.condicion:  # Adquirimos el lock y la condición
            time.sleep(2)  # Simulamos alguna operación
            self.listo = True
            print(f"{threading.current_thread().name} señalando condición.")
            self.condicion.notify()  # Señalamos que la condición se ha cumplido

def main():
    ejemplo = EjemploCondition()

    # Crear un hilo que espera la condición y otro que la señala
    hilo1 = threading.Thread(target=ejemplo.esperar, name="Hilo Espera")
    hilo2 = threading.Thread(target=ejemplo.senalar, name="Hilo Señal")

    hilo1.start()
    hilo2.start()

    hilo1.join()
    hilo2.join()

if __name__ == "__main__":
    main()

File: PracticaVariablesdeCondicion/EjemploCondition.java
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class EjemploCondition {
    private final Lock lock = new ReentrantLock();
    private final Condition condicion = lock.newCondition();
    private boolean listo = false;

    public static void main(String[] args) {
        EjemploCondition ejemplo = new EjemploCondition();

        // Crear un hilo que espera la condición y otro que la senala
        new Thread(() -> ejemplo.esperar()).start();
        new Thread(() -> ejemplo.senalar()).start();
    }

    // Método que espera hasta que se cumpla la condición
    public void esperar() {
        lock.lock();
        try {
            while (!listo) {
                System.out.println(Thread.currentThread().getName() + " esperando...");
                condicion.await();  // Espera a que se cumpla la condición
            }
            System.out.println(Thread.currentThread().getName() + " ha continuado.");
        } catch (InterruptedException e) {
            // Asegúrate de manejar la excepción correctamente
            
        } finally {
            lock.unlock();
        }
    }

    // Método que senala cuando la condición se ha cumplido
    public void senalar() {
        lock.lock();
        try {
            Thread.sleep(2000);  // Simulamos alguna operación
            listo = true;
            System.out.println(Thread.currentThread().getName() + " senalando condicion.");
            condicion.signal();  // Senalamos que la condición se ha cumplido
        } catch (InterruptedException e) {
            // Asegúrate de manejar la excepción correctamente
            
        } finally {
            lock.unlock();
        }
    }
}
